revocation/ocsp/ocsp.go

```diff
@@ -0,0 +1,293 @@
+// Package OCSP provides methods for checking the OCSP revocation status of a
+// certificate chain, as well as errors related to these checks
+package ocsp
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/x509"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/notaryproject/notation-core-go/revocation/base"
+	coreX509 "github.com/notaryproject/notation-core-go/x509"
+	"golang.org/x/crypto/ocsp"
+)
+
+// Options specifies values that are needed to check OCSP revocation
+type Options struct {
+	CertChain   []*x509.Certificate
+	SigningTime time.Time
+	HTTPClient  *http.Client
+}
+
+const (
+	pkixNoCheckOID      string = "1.3.6.1.5.5.7.48.1.5"
+	ocspMaxResponseSize int64  = 20480 //bytes
+)
+
+// CheckStatus checks OCSP based on the passed options and returns an array of
+// base.CertRevocationResult objects that contains the results and error. The
+// length of this array will always be equal to the length of the certificate
+// chain.
+func CheckStatus(opts Options) []*base.CertRevocationResult {
+	result := make([]*base.CertRevocationResult, len(opts.CertChain))
+
+	if len(opts.CertChain) == 0 {
+		return result
+	}
+
+	// Validate cert chain structure
+	if len(opts.CertChain) == 1 {
+		if err := validateIssuer(opts.CertChain[0], opts.CertChain[0]); err != nil {
+			fillResultsWithError(result, base.InvalidChainError{IsInvalidRoot: true, Err: err})
+			return result
+		}
```

Comment:
Sorry I missed this in last revision for `InvalidChainError` can we return just return error i.e. no need to fill error for revocation result as we didnt even checked the revocation status of certs

```
func CheckStatus(opts Options) []*base.CertRevocationResult, error {
```

Reply:
For the first comment about using ValidateCodeSigningCertChain, the issue is related to the leaf assumption for length 1. This would cause it to fail for chains that only contain a root. As a result, we would want to simply check the signature in the same way as it is done in ValidateCodeSigningCertChain without validating the CA/Leaf specific extensions, and then return the NonRevokable if there is no issue (rather than specifically return non-revokable here, it is done automatically further down in the code where it does for other roots)

Reply:
For the second comment about filling InvalidChainError, the main purpose of this is to guarantee the length of results is equal to the length of the cert chain. If we don't want to include this guarantee, we can include a single result with just that error (and corresponding results of Unknown)

Reply:
Makes sense. I can update the tests and change this to always use that validation function